### DISCUSSION TITLE
How to detect when `prisma migrate dev` is needed

### DISCUSSION THREAD

**Discussion by electrovir** [prisma/prisma#16054] [Q&A] [answered]
How can I detect when the schema file has new changes relative to all the saved migrations?

Why:

I would like to automatically detect when `prisma migrate dev` needs to be run (when the schema file has changed but there has not been a migration created for it yet) in a server startup script. I could just naively run `prisma migrate dev` on every startup. However, the database in question's `prisma migrate dev` command takes _a very long time_ so I can't run `prisma migrate dev` on every startup as it will significantly harm the dev experience.

**Comment by austincrim** [comment, marked as answer]
Hey @electrovir 👋,

You should be able to use [`prisma migrate diff`](https://www.prisma.io/docs/reference/api-reference/command-reference#migrate-diff) to build this kind of functionality.

`npx prisma migrate diff --from-schema-datamodel=path/to/schema --to-schema-datasource=path/to/schema` will provide a SQL output if your schema and database are out-of-sync. I'm not sure what the performance of the command will look like for your situation, but definitely something to try.

**Reply by electrovir** [reply]
Ah! That is it!

I had found `migrate diff` but was trying to use `--to-migrations` instead of `--to-schema-datasource`. When I tried to use `--to-migrations` though, it threw `Error: You must pass the --shadow-database-url if you want to diff a migrations directory.` and I had no idea what to do with that (even after much Googling).

In my sample size of just me for this specific database, `prisma migrate diff` is 60x faster than `prisma migrate dev`. So this is much better. Thank you so much!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
